### PR TITLE
Memoize per-application lifecycle replay across scrub buckets

### PR DESCRIPTION
**Pull request by futuroptimist**
## Summary

Phase 2 of the diagram-performance roadmap (see #1198 and
`docs/design/lifecycle-diagram-scrubber-performance-roadmap.md`; this phase is #1194).

Even after Phase 1's bucket-level caching (PR #1199), the *first* visit to any given scrub
bucket still called `projectApp(app, appEvents, isCurrent)` — a dense per-application
event-replay state machine — fresh for **every** included application, even though a typical
adjacent-bucket transition only changes one (or a few) applications' event sets. Every other
application's correct result was byte-identical to the previous bucket's and got needlessly
recomputed.

## Design decision: memoization wrapper, not a state-machine rewrite

Issue #1194 originally framed this as "restructure `projectApp` to update state incrementally"
and flagged it as the highest data-correctness risk in the whole roadmap, since `projectApp` is
covered by ~30 edge-case tests (reopen handling, terminal locking, milestone regression,
same-bucket tie-breaking, supersession, ~15 warning codes).

Investigation found a lower-risk path: `projectApp` is already a **pure function** of
`(app, appEvents, isCurrent)` — it internally re-sorts its input (proven order-independent by an
existing test), event objects have stable identity per bundle (via Phase 1's `prepare()` cache),
and `supersedesEventId` is enforced same-application-only at both the writer (`tracker.js`) and
schema-validation layers (`src/domain/browserApplication.js`) — so one application's included
events can never be affected by another's. This means **the ordered list of included event ids
for `(appId, isCurrent)` is a fully safe, collision-free memoization key**, and wrapping
`projectApp` in a pure cache is provably correct by construction: a cache can only ever return a
value the real function already produced for that exact input — it cannot introduce a new,
independently-fallible computation path the way a hand-rolled incremental reducer could.

All 31 pre-existing tests in `test/web-tracker-lifecycle-projection.test.js` needed **zero
changes** to prove this — direct evidence the wrapper is transparent.

## Changes

- Extended the existing per-bundle cache entry (`lifecycleProjection.js`, from Phase 1) with
  `appPaths`, a `Map<signature, path>` where `signature` is built from `app.id` + `isCurrent` +
  the ordered included event ids (`JSON.stringify`-encoded, collision-safe).
- Added `projectAppCached` as a thin wrapper around `projectApp` and swapped it into
  `projectLifecycleAtUncached`. `projectApp` itself is completely untouched.
- Bounded the cache with a cap sized off `apps.length` (`Math.max(500, apps.length * 6)`) rather
  than a small fixed constant like Phase 1's bucket-level cache — a single full-bundle `"current"`
  bucket visit alone can produce up to `apps.length` fresh signatures, so a small fixed cap would
  thrash constantly.
- Cached path objects are frozen exactly as before (via the existing `deepFreeze(projection)` at
  the end of `projectLifecycleAtUncached`, which is idempotent) — confirmed every consumer of
  `paths` (`makeNodes`, `makeLinks`, `countBy`, the warnings/terminal reductions) is read-only, so
  reusing a cached path object across multiple buckets' `paths` arrays is safe.
- Recorded the actual implemented design (and the busy-indicator deferral, see below) in
  `docs/design/lifecycle-diagram-scrubber-performance-roadmap.md`, which previously still
  described Phase 2 as the unimplemented incremental-reducer approach.

## Tests

- Reference-identity test: an application with no new events between two buckets gets its exact
  cached path object reused; one that does change gets a fresh, different result.
- `isCurrent`-boundary test: an application whose event-id-set is identical at a historical bucket
  and at `"current"` still gets independently correct (non-shared) results, since `isCurrent`
  gates the `status_mismatch` warning.
- LRU-eviction test for the new `appPaths` cache, mirroring Phase 1's pattern.
- Randomized fuzz-equivalence test: a long random-order walk across a large randomized bundle
  (many apps, event types, reopens), compared against a guaranteed-cold computation on a fresh
  content-identical bundle clone for a sample of buckets — asserts deep equality of `paths`,
  `nodes`, `links`, `totals`, `warnings`. Cheap insurance against a bug in signature construction
  specifically (id-ordering, `isCurrent` handling, cap/eviction interaction), as requested by
  #1194, even though the design is provably safe by construction.
- New `staggeredBundle` fixture in the perf test file: the existing `largeBundle` fixture is
  adversarial to this optimization (every application shares identical timestamps, so every dated
  bucket changes every application's event set at once). `staggeredBundle` gives every application
  distinct timestamps so bucket transitions advance roughly one application at a time.
- **Deterministic cache-reuse regression guard** (final form, after two rounds of review — see
  below): over a contiguous window of 50 buckets starting right after all 100 applications have
  entered (so every adjacent-bucket transition changes exactly one application's event set), it
  walks the window on one persistent bundle and counts path references reused vs. recomputed by
  identity, asserting the exact counts this structure implies — 4851 reused / 49 changed / 149
  unique path objects materialized — then walks the same window with a fresh bundle clone per
  call and asserts all 5000 resulting paths are distinct objects (no reuse possible). No timing,
  no threshold, no retries: it's a structural fact, not a wall-clock ratio.
- Every new/changed test was verified to fail against the prior (unmemoized) code with the
  expected symptom, and pass with the fix, by temporarily reverting just the cache wrapper in a
  throwaway copy of `lifecycleProjection.js` (restored before every commit — production file has
  zero diff in every one of these checks).

## Review feedback addressed

Two rounds of automated review on the perf test specifically:

1. Codex found the original evenly-spaced-sample timing benchmark could produce an occasional
   near-miss under timing noise (measured `coldDuration=410ms` against a `477ms` threshold on one
   run) and added ~11s to the suite. First fix: a contiguous post-ramp-up window, a warm-up pass,
   and the median of 7 trials against a 9x threshold — root-caused to the sampling strategy
   diluting the signal, not just insufficient margin.
2. That still wasn't fully host-independent: a follow-up review found two consecutive runs on
   another machine measuring only ~6.08x against the 9x threshold. Any wall-clock ratio, however
   calibrated, is inherently host-dependent. Replaced the whole timing approach with the
   deterministic reference-identity guard described above, which needs no calibration at all.

Also: Copilot flagged the `isCurrent`-boundary test's name as describing the opposite of what it
verifies. Renamed.

## Deferred from #1194's original scope

The busy/loading-indicator item is deferred to Phase 5: the render pipeline is fully synchronous
end-to-end, so a real indicator needs a `requestAnimationFrame`-style deferral point — a separate
architectural change better solved once, alongside the async/Worker offload already planned for
that phase. Confirmed with the user before proceeding.

Closes #1194.

## Test plan

- [x] `npx vitest run test/web-tracker-lifecycle-projection.test.js test/web-tracker-lifecycle-diagram-performance.test.js test/web-tracker-lifecycle-diagram.test.js` — 72 tests pass (35 + 3 + 34), including all 31 pre-existing edge-case tests unmodified.
- [x] `npm run test:ci` (full suite, 1270 tests total) — 1263 pass, 1 skipped, 6 fail. All 6 failures are pre-existing and unrelated to any file this PR touches: 5 are `test/cli.test.js` failures caused by `better-sqlite3 unavailable` (native-module binding issue in this sandbox), and 1 is the same already-documented, environment-flaky `lifecycle-diagram-layout` "fan-in fixtures" test noted as unrelated in PR #1199.
- [x] The final deterministic perf test: verified it fails against a temporarily-reverted (unmemoized) baseline and passes reliably across 5 repeated full-file `vitest` runs — no flakiness, since it makes no timing assertions.
- [x] `npm run typecheck`, `eslint`, and `prettier --check` — clean.
- [x] Manually verified in a running dev server against real IndexedDB data (200 seeded applications/1600 events): Diagram tab renders correctly, scrubbing works, "Return to current" snaps back correctly, no console errors.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: test/web-tracker-lifecycle-projection.test.js [hunk 1]
```diff
@@ -135,6 +135,166 @@ describe("lifecycle projection", () => {
...
+    // Application "a" gets no new events between bucketA and bucketB, so
+    // its included-event-id-set is identical — the cached path object must
+    // be reused verbatim rather than recomputed.
+    expect(pathAt(bucketB, "a")).toBe(pathAt(bucketA, "a"));
+    // Application "b" gains an event at bucketB, so its path must differ.
+    expect(pathAt(bucketB, "b")).not.toBe(pathAt(bucketA, "b"));
+    expect(pathAt(bucketB, "b").milestones).toContain("recruiter_screen");
+  });
+
+  it("keeps isCurrent out of the app-path cache key from colliding with historical buckets", () => {
```
[the remark this answers is not in the document]
Comment (futuroptimist) [reply test/web-tracker-lifecycle-projection.test.js]: Fixed — renamed to "includes isCurrent in the app-path cache key to avoid historical-bucket collisions", which describes what the test actually verifies. See 2c8ea4c.

File: test/web-tracker-lifecycle-diagram-performance.test.js [hunk 2]
```diff
@@ -221,4 +270,43 @@ describe("lifecycle diagram large-data rendering", () => {
...
+
+    const persistentBundle = staggeredBundle(appCount);
+    const warmStart = performance.now();
+    for (const id of bucketIds) projectLifecycleAt(persistentBundle, id);
+    const warmDuration = performance.now() - warmStart;
+
+    const coldStart = performance.now();
+    for (const id of bucketIds) {
+      const freshClone = staggeredBundle(appCount);
+      projectLifecycleAt(freshClone, id);
```
[the remark this answers is not in the document]
Comment (futuroptimist) [reply test/web-tracker-lifecycle-diagram-performance.test.js]: Good catch. Sampled ~50 evenly-spaced buckets instead of walking all ~800, and recalibrated the threshold against both the fixed and reverted code (confirmed ~6.7x ratio with the fix vs ~3.5x without it, both stable across repeated runs) — a 5x threshold cleanly separates them with margin. Total test time is now well under 200ms instead of ~10.8s. See 2c8ea4c.

[the remark this answers is not in the document]
Comment (futuroptimist) [reply test/web-tracker-lifecycle-diagram-performance.test.js]: Root-caused and fixed in 77a9804. The evenly-spaced sample mixed in the early ramp-up region (few included applications, noise-dominated timings), which could produce an occasional near-miss. Switched to a contiguous window starting right after every application has entered (allBucketIds.slice(appCount, appCount + windowSize)), added a disposable warm-up pass, and now take the median of 7 independent trials instead of one measurement. Verified stable ~13-14x (fixed) vs ~6.5-7x (unmemoized, checked via a throwaway copy — production file has zero diff) across 5+ repeated process runs and 5 repeated full-file vitest runs; a fixed 9x threshold sits comfortably clear of both. Also confirmed the target test still fails reliably (3/3 repeats) against the unmemoized baseline.

[the remark this answers is not in the document]
Comment (futuroptimist) [reply test/web-tracker-lifecycle-diagram-performance.test.js]: Agreed — replaced the wall-clock benchmark entirely with a deterministic reference-identity guard in 1eb119a, per the diagnostic in the merge-readiness note. Over the same 100-application/50-bucket contiguous window, it counts reused-vs-changed path references across adjacent buckets (asserts exactly 4851 reused / 49 changed / 149 unique objects for the persistent walk, 5000 unique for the fully-cold walk) — independently reproduced those exact numbers before implementing. No timing, no threshold, no retries. Confirmed it fails deterministically (not probabilistically) against a temporarily unmemoized baseline in a throwaway copy of lifecycleProjection.js, and passes reliably across 5 repeated full-file runs plus test:ci.
